### ISSUE TITLE
nix-build ignores SIGPIPE

### ISSUE THREAD

**Issue by bjornfor** [NixOS/nix#2653]
I'm running nix-build in a gitolite VREF, to detect build problems at git push time. The problem is if the client aborts the push with "ctrl-c", nix-build still runs until completion.

To reproduce:

```
$ nix-build slow.nix |& sleep 1
```

where slow.nix is something like:

```nix
with import <nixpkgs> {};

stdenv.mkDerivation {
  name = "slow-build";
  buildCommand = ''
    total=3600
    for i in $(seq "$total"); do
        echo "$i/$total"
        sleep 1
    done
    touch "$out"
  '';
}
```

I expect nix-build to abort after 1 second, freeing resources, but instead it completes the build (>> 1 second).

Tested nix v2.1.3 and v2.2.

**Comment by bjornfor** [comment]
The above use case gets fixed with this tiny patch:

```patch
diff --git a/src/libutil/util.cc b/src/libutil/util.cc
index ce50334e..f18104ef 100644
--- a/src/libutil/util.cc
+++ b/src/libutil/util.cc
@@ -1470,7 +1470,6 @@ void startSignalHandlerThread()
     sigaddset(&set, SIGINT);
     sigaddset(&set, SIGTERM);
     sigaddset(&set, SIGHUP);
-    sigaddset(&set, SIGPIPE);
     sigaddset(&set, SIGWINCH);
     if (pthread_sigmask(SIG_BLOCK, &set, nullptr))
         throw SysError("blocking signals");
```

```
$ time ./result/bin/nix-build slow.nix |& sleep 1

real	0m1.739s
user	0m0.229s
sys	0m0.042s
```

However, when looking at git log it seems using SIGPIPE had something to do with nix-shell. Like, the last commit to touch that SIGPIPE line was commit 7a65b2470 by @edolstra with the message:

```
Restore default signal handling in child processes

In particular, this fixes Ctrl-C in nix-shell sessions.
```

But when pressing Ctrl-C in nix-shell (with the above patch), I'm not seeing anything out of the ordinary:
 
```
$ ./result/bin/nix-shell -A coreutils '<nixpkgs>'
[nix-shell:~]$ 
[nix-shell:~]$ ^C
[nix-shell:~]$ 
```

**Comment by vcunat** [comment]
I suspect it can't be solved this way.  In general one can get SIGPIPE for various reasons, e.g. when using TCP.   Nix certainly uses TCP, but I don't know details about threads/processes in nix for this.  The cleanest way will probably be to check for errors during write of the output.

**Comment by bjornfor** [comment]
Ok. I didn't know about TCP + SIGPIPE. Apparently that case can be handled with

```
int n = 1;
setsockopt(thesocket, SOL_SOCKET, SO_NOSIGPIPE, &n, sizeof(n))
```

on linux (it's not portable). So you're probably right that it's best to check the status of writes. Not sure where to begin though...

**Comment by vcunat** [comment]
Yeah, I also didn't have idea that SIGPIPE extends to network streams until we started getting them (at work).

**Comment by bjornfor** [comment]
Following the writes lead me to this code, in src/libutil/logging.cc:

```cpp
void writeToStderr(const string & s)
{
    try {
        writeFull(STDERR_FILENO, s, false);
    } catch (SysError & e) {
        /* Ignore failing writes to stderr.  We need to ignore write
           errors to ensure that cleanup code that logs to stderr runs
           to completion if the other side of stderr has been closed
           unexpectedly. */
    }
}
```

I think that leaves these options:
* Add optional argument to writeToStderr, e.g. bool allowErrors.
* Add global variable, e.g. bool teardownInProgress that, when set, ignores errors.
* Use non-portable socket API to disable SIGPIPE on TCP sockets.

**Comment by bjornfor** [comment]
Still happens.

I generally don't want to ping the stale bot to un-stale, because I feel like it just adds to the noise. I didn't expect this issue to be closed. I'm happy to see people are taking action and hopefully stalebot closing tasks will be stopped (or removing the stale bot alltogether, as I think it's just creating noise).

**Comment by bjornfor** [comment]
@edolstra: Please re-open.
